kyaml/kyaml.go

```diff
@@ -0,0 +1,705 @@
+	mapTag       = "!!map"
+	nullTag      = "!!null"
+	binaryTag    = "!!binary"
+	mergeTag     = "!!merge"
+)
+
+type flagMask uint64
+
+const (
+	flagsNone     flagMask = 0
+	flagLazyQuote flagMask = 0x01
+	flagCompact   flagMask = 0x02
+)
+
+// renderDocument processes a YAML document node, rendering it to the output.
+// This function assumes that the output "cursor" is positioned at the start of
+// the document and should always emit a final newline.
+func (ky *Encoder) renderDocument(doc *yaml.Node, indent int, _ flagMask, out io.Writer) error {
+	if len(doc.Content) > 1 {
+		return fmt.Errorf("kyaml internal error: line %d: document has more than one root node (%d)", doc.Line, len(doc.Content))
+	}
+	if indent != 0 {
+		return fmt.Errorf("kyaml internal error: line %d: document non-zero indent (%d)", doc.Line, indent)
+	}
+
+	// For document nodes, the cursor is assumed to be ready to render.
+	if len(doc.HeadComment) > 0 {
+		ky.renderComments(doc.HeadComment, indent, out)
+		fmt.Fprint(out, "\n")
+	}
+	child := doc.Content[0]
+	if len(child.HeadComment) > 0 {
+		ky.renderComments(child.HeadComment, indent, out)
+		fmt.Fprint(out, "\n")
+	}
+	if err := ky.renderNode(child, indent, flagsNone, out); err != nil {
+		return err
+	}
+	if len(child.LineComment) > 0 {
+		ky.renderComments(" "+child.LineComment, 0, out)
+	}
+	fmt.Fprint(out, "\n")
+	if len(child.FootComment) > 0 {
+		ky.renderComments(child.FootComment, indent, out)
+		fmt.Fprint(out, "\n")
+	}
+	if len(doc.LineComment) > 0 {
+		ky.renderComments(" "+doc.LineComment, 0, out)
+		fmt.Fprint(out, "\n")
+	}
+	if len(doc.FootComment) > 0 {
+		ky.renderComments(doc.FootComment, indent, out)
+		fmt.Fprint(out, "\n")
+	}
+	return nil
+}
+
+// renderScalar processes a YAML scalar node, rendering it to the output.  This
+// DOES NOT render a trailing newline or head/line/foot comments, as those
+// require the parent context.
+func (ky *Encoder) renderScalar(node *yaml.Node, indent int, flags flagMask, out io.Writer) error {
+	switch node.Tag {
+	case intTag, floatTag, boolTag, nullTag:
+		fmt.Fprint(out, node.Value)
+	case strTag:
+		return ky.renderString(node.Value, indent+1, flags, out)
+	case timestampTag:
+		return ky.renderString(node.Value, indent+1, flagsNone, out)
+	default:
+		return fmt.Errorf("kyaml internal error: line %d: unknown tag %q on scalar node %q", node.Line, node.Tag, node.Value)
+	}
+	return nil
+}
+
+const kyamlFoldStr = "\\\n"
+
+// renderString processes a string (either single-line or multi-line),
+// rendering it to the output.  This DOES NOT render a trailing newline.
+func (ky *Encoder) renderString(val string, indent int, flags flagMask, out io.Writer) error {
+	lazyQuote := flags&flagLazyQuote != 0
+	compact := flags&flagCompact != 0
+
+	// If no newlines, just use standard Go quoting.
+	if compact || !strings.Contains(val, "\n") {
+		if lazyQuote && !needsQuotes(val) {
+			fmt.Fprint(out, val)
+		} else {
+			fmt.Fprint(out, strconv.Quote(val))
```

Comment:
is standard go quoting 100% compatible with yaml string escaping?

Reply:
fair question.  I will have to evaluate that and get back

Reply:
Should be compatible. 

`\t` `\n`, `\xFF` `\u0100` escape style are all in "Example 5.13 Escaped Characters" in YAML 1.2.2 spec. 

I also see these in 1.1 "Example 5.14. Escaped Characters" and even the 1.0 draft "4.6.9. Escaping".

Reply:
If I read the YAML spec correctly, `'\x00'` should escape in YAML as `"\0"` but Go converts it to `"\x00"` - so I probably have to decode manually.

Reply:
Actually, I have no idea how to read that table - it includes space and slash, which clearly are not actually escaped

Reply:
> If I read the YAML spec correctly, '\x00' should escape in YAML as "\0" but Go converts it to "\x00" - so I probably have to decode manually.

I don't think `\x00` is *invalid* in yaml, but `\0` would probably be canonical. https://go.dev/play/p/a6_gobBEp0k